src/assignments/tech-jobs-oo.rst

```diff
@@ -1,6 +1,586 @@
+   b. Select the defaults in all the other panes.
+
+TechJobs (Object-Oriented Edition)
+-----------------------------------
+
+Sally has gotten the ball rolling by adding a ``Job`` class, along with classes
+to represent the individual properties of a job: ``Employer``, ``Location``,
+``PositionType``, and ``CoreCompetency``. She completed the ``Employer`` class,
+and she left you the task of filling in the others.
+
+As the team gets closer to deploying the app---and abandoning the test data
+they’ve been using---they’ll want an easy way to add and remove jobs via a
+user interface. Before that, however, you need to finish shifting the project
+to an object-oriented design.
+
+Why Change to Object-Oriented?
+-------------------------------
+
+Working with data stored as strings in HashMaps and ArrayLists isn’t a good
+long-term solution, for reasons that we point out below.
+
+The ``Job`` class introduces an object-oriented design to the application. It
+contains all of the fields you used in the console version of TechJobs:
+``name``, ``employer``, ``location``, ``positionType``, ``coreCompetency``.
+There’s also an ``id`` field which will be used to uniquely identify ``Job``
+objects.
+
+The main difference between the object representation of a job and the
+string-based representation is that the values of ``employer``, ``location``,
+and the other non-ID fields are no longer strings. Instead, they are classes of
+their own.
+
+Job Fields
+^^^^^^^^^^^
+
+Open the ``Job`` class file. You’ll see the following fields (among others):
+
+.. sourcecode:: java
+   :linenos:
+
+   private String name;
+   private Employer employer;
+   private Location location;
+   private PositionType positionType;
+   private CoreCompetency coreCompetency;
+
+Of these, only ``name`` is a string. Sally created classes to represent each of
+the other properties. These classes---``Employer``, ``Location``,
+``CoreCompetency``, ``PositionType``---have ``value`` and ``id`` fields.
+
+So, for example, if you had a ``Job`` instance, you could get the name of the
+employer this way:
+
+.. sourcecode:: java
+
+   // job is an instance of Job
+   String employerName = job.getEmployer().getValue();
+
+Additionally, the ``toString()`` method of the ``Employer`` class is set up to
+return the ``value`` field. Thus, using one of these objects in another string
+context like ``System.out.println`` will print the data stored in ``value``.
+
+.. sourcecode:: java
+
+   // Prints the name of the employer
+   System.out.println(job.getEmployer);
+
+Why do we go to all of this trouble when we could store this job-related data
+as strings? There are a couple of reasons.
+
+Eliminate Duplication of Data
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In our app, we have multiple jobs that have the same value in a given field.
+For example, there are multiple jobs with position type “Web - Full Stack”, and
+there are different jobs with the same employer. If we store the values of
```

Comment:
I suggest rephrasing this as "...each employer may have multiple jobs." This emphasizes the one-to-many aspect of the relationship.

Reply:
Fixed.  Thanks.